### ISSUE TITLE
UART not working after boot_code

### ISSUE THREAD

**Issue by UnaiGalicia** [quangdaovu/arty_pulpino#2]
Hi,

First of all, thank you very much for porting pulpino to the Arty board. That really saves time.

Here's my issue: I modified the boot_code.c file and transferred the SV file to pulpino. I use the UART in that file with the uart_send function, and it works perfectly. However, after the boot_code loads the program in the instruction RAM, that program doesn't produce any output using uart_send or printf. 

I tried that same program in the Zynq and it worked perfectly, I wonder if you could help me out here.

Thank you again!

**Comment by quangdaovu** [comment]
Hi,

It has been long time since the last time I worked with PULPino, therefore I can barely remember the whole details.
Can you describe your procedure in more details? It is better to be step by step. Doing that will be a big help for me to find out what is the issue.

**Comment by UnaiGalicia** [comment]
Sure. I changed a little the boot_code.c, where the program is read from the flash memory and written into the instruction memory. There, I used the uart_send function (just like you did) to print some lines, and it worked fine, I get everything that I send in my terminal. Then, after writing from flash, I made the program jump to the instruction RAM.

In the program that I loaded into the instruction RAM, I included the same libraries as in the boot_code.c file, and I used the same uart_send function. However, I don't get anything in the terminal. I then included the following in the program loaded into RAM:

`while(1)
    uart_send("Hello\n, 6);`

And the Arty board's UART TX LED lights up continuously, but I still don't get anything in the terminal. I even tried to use the uart_set_cfg(0, 26) as you use in the boot_code.c, but that didn't work either.

I ran that same program that I loaded in the instruction RAM in the original PULPino using a Zedboard and the uart_Send command works just fine. So, I don't know if I'm missing something.

**Comment by quangdaovu** [comment]
> Then, after writing from flash, I made the program jump to the instruction RAM.

Do you mean "reading from FLASH"?
Let me conclude your process:
1. Write a "hello world" program that has
`while(1) uart_send("Hello\n, 6);`
2. Program Instruction vector to 0xFF0000 and Data vector to 0xFF8000 of Arty's SPI FLASH.
3. Using option 2 of your boot_code (which has some extra uart_send functions).
4. After option 2 finished, using option 1?

I would recommend you to insert another step in between 3 and 4, which is assert reset for the core (by pushing RST button, I don't remember exactly, but it is there). The root cause is that step 3 will try to overwrite boot_code own stack region, which resulted in corrupted boot_code. Reset the core will not reset memories, boot_code will surely overwrite *some* data in its stack region, but I don't think it is important (since your program is very small, it will not go up to boot_code stack region).

Regarding your custom program, did you tested it with JTAG?

Also, `uart_set_cfg(0, 26)` is not needed because it has been called in crt0.riscv.S already, and you can use `printf` instead of `uart_send`.

**Comment by UnaiGalicia** [comment]
So, correct me if i'm wrong:
1. Running the boot_code that transfers the program from flash to memories
2. From the boot_code, somehow reset the core (I found in the pulpino datasheet that the Interrupt Vector Table contains a  reset interrupt @ address 0x80 )
3. System should continue reading from the RAM memories

Wouldn't the core start again from the boot_code (step 2) and form a loop?

**Comment by quangdaovu** [comment]
Note that current boot code is not really in finished state yet. I had to move to another project at that time.

You can safely remove that "reset" step by using only 28KB of data memory.

If you take a look at link.boot.ld, stack of boot_code is the last 4KB of data memory. Therefore, you can modify link.common.ld to use only 28KB from start of data memory to the beginning of this 4KB region (i.e. from 0x100000 to 0x00106000) --> no overlap at all.

Also, you will have to modify boot_code to only read 28KB data from SPI FLASH [here](https://github.com/quangdaovu/arty_pulpino/blob/49b124d5e89b617d407854411b8f562dcc773e52/sw/apps/boot_code/boot_code.c#LL121C14-L121C14).

**Comment by UnaiGalicia** [comment]
So, you mean that when loading the contents of the data memory, the stack of the boot code is overwritten? And that the data memory should only occupy from 0x100000 to 0x106000 so that stack isn't overwritten? But, aren't those values the default ones in link.common.ld ?

Then when my program jumps from the boot code to the instruction memory's content, why is the program executed correctly until the uart_send function's execution?

I also tried to modify the boot_code.c so that instead of reading from flash, it jumped directly to the instruction memory, having stored in the memories the correspondent coe files (with a simple program that prints something using uart_send), but still I get no output in the terminal. However, if I use the content of the uart_sendchar function in a while loop in the loaded program instead of uart_send:

` *(volatile unsigned int*)(UART_REG_THR) = 'c'; `

I get the character 'c' in the terminal. I don't really understand what is happening here, what do you think?

Sorry if my explanations are confusing, I know that remembering all this is probably difficult for you.

**Comment by quangdaovu** [comment]
> So, you mean that when loading the contents of the data memory, the stack of the boot code is overwritten? And that the data memory should only occupy from 0x100000 to 0x106000 so that stack isn't overwritten? But, aren't those values the default ones in link.common.ld ?

Yes, that's what I mean. Currently stack region of boot_code & normal app are overlapped. Most of the time it would not be the problem, but in boot_code.c line 112 to 132, that stack region is overwritten by the content of SPI FLASH, thus might leads to corrupted stack. 

>Then when my program jumps from the boot code to the instruction memory's content, why is the program executed correctly until the uart_send function's execution?

> I also tried to modify the boot_code.c so that instead of reading from flash, it jumped directly to the instruction memory, having stored in the memories the correspondent coe files (with a simple program that prints something using uart_send), but still I get no output in the terminal. However, if I use the content of the uart_sendchar function in a while loop in the loaded program instead of uart_send:

> *(volatile unsigned int*)(UART_REG_THR) = 'c';

> I get the character 'c' in the terminal. I don't really understand what is happening here, what do you think?

Does `uart_sendchar` work? If `uart_sendchar` does not work but above line works then the the only reason is UART FIFO not empty. 

`*(volatile unsigned int*)(UART_REG_FCR) = 0xA7; //enables 16byte FIFO and clear FIFOs`

This might help.

**Comment by UnaiGalicia** [comment]
I tried to execute

```
*(volatile unsigned int*)(UART_REG_FCR) = 0xA7; //enables 16byte FIFO and clear FIFOs
uart_sendchar('c');
```
But I didn't get any output. However, by executing the contents of `uart_set_cfg(0,26)` without calling the function, like this:

```
CGREG |= (1 << CGUART); // don't clock gate UART
*(volatile unsigned int*)(UART_REG_LCR) = 0x83; //sets 8N1 and set DLAB to 1
*(volatile unsigned int*)(UART_REG_DLM) = (26 >> 8) & 0xFF;
*(volatile unsigned int*)(UART_REG_DLL) =  26       & 0xFF;
*(volatile unsigned int*)(UART_REG_FCR) = 0xA7; //enables 16byte FIFO and clear FIFOs
*(volatile unsigned int*)(UART_REG_LCR) = 0x03; //sets 8N1 and set DLAB to 0
*(volatile unsigned int*)(UART_REG_IER) = ((*(volatile unsigned int*)(UART_REG_IER)) & 0xF0) | 0x02; 
```
And then,

```
uart_sendchar('y');	//XXX
printf("Test\n");
```

I get at the output both the character and the string. Maybe more than one UART control registers need to be reset? Still, all this looks strange to me.

**Comment by UnaiGalicia** [comment]
Hi,

I finally found out that my boot code copied the contents of the data memory from FLASH into the data RAM, thus it was rewriting the stack of the boot_code program, as you pointed out. My solution, however, was to replace 

```
jump_and_start = (void(*)(void))INSTR_RAM_START_ADDR;
jump_and_start();
```

with

`__asm__ __volatile__("jalr x0, x0, 0x80" );`

This way the content of the instruction is in the ROM, thus it is not rewritten when dumping the contents of the FLASH into the data RAM. I don't mean that what you initially wrote is wrong, but with the use that I gave to the boot_code it doesn't work. Thank you for your time, quangdaovu.

By the way, when you developed this port to the Arty board, or when using the pulpino+pulpemu, did you try the freertos application? I am having some issues as the program freezes, probably due to incorrect task stack usage. Many thanks again.
